synapse/handlers/auth.py

```diff
@@ -2285,3 +2300,45 @@ async def on_logged_out(
             except Exception as e:
                 logger.warning("Failed to run module API callback %s: %s", callback, e)
                 continue
+
+    async def get_username_for_registration(
+        self,
+        uia_results: JsonDict,
+        params: JsonDict,
+    ) -> Optional[str]:
+        """Defines the username to use when registering the user, using the credentials
+        and parameters provided during the UIA flow.
+
+        Stops at the first callback that returns a string.
+
+        Args:
+            uia_results: The credentials provided during the UIA flow.
+            params: The parameters provided by the registration request.
+
+        Returns:
+            The localpart to use when registering this user, or None if no module
+            returned a localpart.
+        """
+        for callback in self.get_username_for_registration_callbacks:
+            try:
+                res = await callback(uia_results, params)
+
+                if isinstance(res, str):
+                    return res
+                elif res is not None:
+                    # mypy complains that this line is unreachable because it assumes the
+                    # data returned by the module fits the expected type. We just want
+                    # to make sure this is the case.
+                    logger.warning(  # type: ignore[unreachable]
+                        "Ignoring non-string value returned by module in "
```

Comment:
Should this say which module returned the bad value? I guess that's not easily available.

Reply:
I've reworded this warning, and made it log the repr for the callback, which looks like this:

```
2022-01-26 11:50:23,425 - synapse.handlers.auth - 2332 - WARNING - POST-17 - Ignoring non-string value returned by get_username_for_registration callback <bound method UsernameFromThreepid.set_username_from_threepid of <username_from_threepid.UsernameFromThreepid object at 0x7fa77ea42020>>: [...]
```

which might be a bit too verbose but better this than not verbose enough.